docs/codeql/codeql-language-guides/customizing-library-models-for-javascript.rst

```diff
@@ -0,0 +1,474 @@
+
+Customizing Library Models for JavaScript
+=========================================
+
+.. include:: ../reusables/beta-note-customizing-library-models.rst
+
+The JavaScript analysis can be customized by adding library models in data extension files.
+
+A data extension for JavaScript is a YAML file of form:
+
+.. code-block:: yaml
+
+  extensions:
+    - addsTo:
+        pack: codeql/javascript-all
+        extensible: <name of extension point>
+      data:
+        - <tuple1>
+        - <tuple2>
+        - ...
+
+The data extension can contribute to the following extension points:
+
+- **sourceModel**\(type, path, kind)
+- **sinkModel**\(type, path, kind)
+- **typeModel**\(type1, type2, path)
+- **summaryModel**\(type, path, input, output, kind)
+
+TODO: mention how to actually load the data extension and/or link to page about data extensions.
+
+We'll explain how to use these using a few examples, and provide some reference material at the end of this article.
+
+Example: Taint sink in the 'execa' package
+------------------------------------------
+
+In this example, we'll show how to add the argument passed to **execa** below as a command-line injection sink:
+
+.. code-block:: js
+
+  import { shell } from "execa";
+  shell(cmd); // <-- add 'cmd' as a taint sink
+
+This sink is already recognized by the CodeQL JS analysis, but for the sake of this example we'll show how it could be added as an extension.
+
+This can be achieved with the following data extension:
+
+.. code-block:: yaml
+
+  extensions:
+    - addsTo:
+        pack: codeql/javascript-all
+        extensible: sinkModel
+      data:
+        - ["execa", "Member[shell].Argument[0]", "command-line-injection"]
+
+To break this down:
+
+- Since we're adding a new sink, we add a tuple to the **sinkModel** extension point.
+- The first column, **"execa"**, identifies a set of values from which to begin the search for the sink.
+  The string **"execa"** means we start at the places where the codebase imports the NPM package **execa**.
+- The second column is an access path that is evaluated from left to right, starting at the values that were identified by the first column.
+
+  - **Member[shell]** selects accesses to the **shell** member of the **execa** package.
+  - **Argument[0]** selects the first argument to calls to that member.
+
+- **command-line-injection** indicates that this is considered a sink for the command injection query.
+
+Example: Taint sources from window 'message' events
+---------------------------------------------------
+
+In this example, we'll show how the **event.data** expression below could be marked as a remote flow source:
+
+.. code-block:: js
+
+  window.addEventListener("message", function (event) {
+    let data = event.data; // <-- add 'event.data' as a taint source
+  });
+
+This source is already known by the CodeQL JS analysis, but we'll show how it could be added as an extension. This can be achieved with the following extension:
```

Comment:
```suggestion
Note that this sink is already added to the CodeQL JavaScript analysis, but for this example, you could use the following data extension:
```

Similar point to above re alignment with Java content

Reply:
Fixed (see https://github.com/github/codeql/pull/11615#discussion_r1147591771)